README.md
The SleepCycles package
================

<!-- README.md is generated from README.Rmd. Please edit that file -->

## Package and main function

The SleepCycles package and its main function `SleepCycles()` is
designed to detect sleep cycles and their respective NREM and REM parts
(called (N)REM periods) from data that has been sleep staged according
to AASM criteria. Additionally, each (N)REM part is split into
percentiles. The function results in a plot visualising the results and
creates a text file so the results can be used for further processing.

### Detection of sleep cycles

Sleep cycles are largely detected according to the originally proposed
criteria by Feinberg & Floyd (1979) and as described in Blume & Cajochen
(2021) . NREM periods are periods starting with N1 (default) or N2 at
the beginning of the night and W or another NREM stage following a REM
period. NREMPs have a minimal duration minimal duration of 15min (can
include W, up to &lt;5min REM, except for the first REMP, for which
there is no minimum duration criterion). REM following a NREM period
always represents a potential REM period, however any REMP must be at
least 5min long (except the first REMP, for which no minimum duration
criterion is applied). If a NREMP exceeds 120min in duration (excl.
wake), it can be split into 2 parts. The new cycle starts with the first
N3 episode following a phase (&gt;12min) with any other stage than N3,
that is a lightening of sleep (cf. Rudzik et al., 2020; Jenni et al.,
2004; Kurth et al., 2010). The function makes suggestions where
splitting could be done according to these criteria and visualises the
potential splitting points on top of a hypnogram. The user can then
interactively choose where to split the NREMP. However, the code also
offers the possibility to provide a numeric value for an epoch at which
to split or you can also decide to not split at all. A combination of a
NREMP and the following REMP represents one sleep cycle, except for the
case when a NREMP is split. In this case, the first of the two resulting
NREMPs represents a sleep cycle (without REM).

### Requirements

The function requires any sleep staging results file with a column, in
which the sleep stages are coded in the usual numeric 0,1,2,3,5 (i.e.,
W, N1, N2, N3, REM) pattern (i.e., a numeric vector). The user can
define other integers to be handled as W or N3 (i.e. in the case
stagings were done according to the Rechtschaffen & Kales criteria
including S3 and S4). The presence of further columns, e.g. a ‘time’
column, is not an issue. Staging must be in 30s epochs. Besides text
files, the `SleepCycles()` function can also handle csv and marker files
for the Brain Vision Analyzer. The input file type can be indicated with
the `filetype` argument (`filetype = "txt"` (default) or
`filetype = "csv"` or `filetype = "vmrk"`).

#### Details

Besides sleep cycles (NREM-REM), the result also splits the NREM and REM
parts of each cycle in percentiles. In case the length of a period is
not divisible by 10 (e.g., 203 epochs), one epoch is added to
percentiles in a randomized fashion to reach the correct length of a
period (here: 7 percentiles of 20 epochs, 3 of 21 epochs).

The code offers to choose whether incomplete periods should be removed
at the end of the night (argument `rm_incomplete_period`, default =
FALSE). Incomplete periods are defined by periods that are followed by
&lt;5min NREM or W (e.g. because a participant is woken up).

Although this is not encouraged, for some participants it may be
necessary to decrease the minimum duration of REM from 5min to 4 or
4.5min as otherwise a seemingly ‘clear’ REM period is skipped. While the
default length of REMPs is 10 segments, it can be decreased.

The user can either process all files in a given directory (default) or
specific files by specifying a vector of files (argument `files`).

By default, the function produces and saves a plot for visual inspection
of the results (argument `plot`, default = TRUE).

#### Arguments in SleepCycle function

`p` character vector indicating the directory containing the sleep
staging files  
`sleepstart` character vector indicating whether the first NREMP at the
beginning of the night should start with N1 or N2. Default: N1 `files`
numeric vector indicating which files in ‘p’ to process. Default: NA  
`filetype` character indicating file type of the files containing the
sleep staging results. Can be “txt” (default) or “csv” or “vmrk” (i.e.,
marker files for Brain Vision Analyzer Software).  
`treat_as_W` numeric vector indicating which values should be treated as
‘wake’. Default: NA  
`treat_as_N3` numeric vector indicating which values should be treated
as ‘N3’. Default: NA  
`rm_incomplete_period` logical: should incomplete periods at the end of
the night be removed? Default: F.  
`plot` logical: should a plot for the result of the detection procedure
be generated and saved? Default: T.  
`REMP_length` numeric value specifying the minimum duration of a REM
period. Default is 10 segments (i.e. 5 minutes). Decreasing the min.
length is not encouraged and should only be done following careful
consideration

### Worked example

First, we install and load the package if we haven’t done so.

``` r
## First, we save your current workspace
save.image(file=paste(tempdir(), "currsession.RData", sep = "/"))
## make sure you start with a clean session.
rm(list = ls(all = TRUE))
install.packages("SleepCycles", repos = "http://cran.us.r-project.org")
```

Then, we are ready to use the package on our data set. Note that in the
`sleepstages2` data set, the first NREM period exceeds 120 minutes.
Thus, the code attempts to split this NREM period.

The text file has a header, thus, when asked whether it has a header
file, type `y`. Columns are separated by comma, thus type `,` when
prompted. Of course, if we had several of these files in our directory,
they would all have to have the same pattern.

##### Create directory & save data file

First, we load the `sleepstages2` file that comes with the package and
create a directory, where we save it.

``` r
library(SleepCycles)
data(sleepstages2)

## save current working directory so we can reset this later.
olddir <- getwd()

## create a new directory in the temporary directory (don't worry, it will automatically be deleted  
## when you restart your computer)
newdir <- file.path(tempdir(),"SleepCycles_exmpl2")
dir.create(newdir, showWarnings = FALSE)

## write the sleepstages2 file to this new directory
write.table(sleepstages2, file = paste(newdir, "sleepstages2.txt", sep = "/"),
row.names=FALSE, col.names = TRUE, quote = FALSE, sep = ",")
```

##### Run the detection

Then, we apply the actual `SleepCycles` function. The file contains
column names in a header and columns are separated with a comma. When we
are prompted, we have to decide where we want to split the data, either
at the first or the second suggested location. I would suggest selecting
the fist, so when prompted, we simply type `1`.

``` r
SleepCycles::SleepCycles(newdir, filetype = "txt")

## We again load the workspace image from before the code above was executed
save.image(file=paste(tempdir(), "currsession.RData", sep = "/"))

## we set the directory back to the one we were using before as we were just working in the  
## temp directory.
setwd(olddir)
```


NEWS.md
The SleepCycles package
================

<!-- NEWS.md is generated from NEWS.Rmd. Please edit that file -->

## News

### SleepCycles 1.1.4 (2021-22-09)

Solved an issue that could result in problems with vignette building.

### SleepCycles 1.1.3 (2021-11-08)

Code in the Readme file that violated CRAN policy has been replaced.

### SleepCycles 1.1.2 (2021-25-04)

Enhancement: now also files in which a sleep stage (e.g. N3) is missing
can be read in.

### SleepCycles 1.1.1 (2021-29-03)

Updated the reference and doi to the companion paper by Blume &
Cajochen. Created a CITATION file for the publication.

### SleepCycles 1.1.0 (2021-18-02)

Bug fix: the load\_prep.R function resulted in an error when input files
had less than two columns, this has been fixed. Bug fix: in the unlikely
event that there is only one REM sleep episode, the code resulted in an
error. This has been fixed. Enhancement: the user can now also specify
csv files as input files to the SleepCycles() function. Enhancement: the
user can now specify in the ‘sleepstart’ argument whether the first
NREMP of a night should start with N1 (default) or N2.

### SleepCycles 1.0.1 (2020-02-12)

No changes since version 0.8. This is the first version available on
CRAN.

### SleepCycles 1.0.0 (2020-24-11)

No changes since version 0.8. This is the first version submitted to
CRAN.

### SleepCycles 0.8 (2020-18-11)

First NEWS update. All changes will be documented in this file from now
on. This is also the first version available on CRAN.


BLUMEC_ALocGlo_SleepCycles.R
##################################################################################################################################
## Code written to analyse sleep cycles from SIESTA Scoring Data    
##                                                                      
## Author: CBlume                                                       
## Date: 22 07 20
## Version: 1.0
##
## The function requires any sleep staging file with a column named "Description", in which the sleep stages are coded 
## in the usual 0,1,2,3,5 (i.e., W, N1, N2, N3, REM) pattern. Staging must be in 30s epochs. Besides text files, it 
## can also handle marker files for the Brain Vision Analyzer (filetype = "txt" (default), "vmrk", or "vpd").
##
## Sleep cycles are largely defined according to the originally proposed criteria by Feinberg & Floyd (1979). 
## NREM periods are periods starting with N1 with a minimal duration of 15min (can include W, but not REM). 
## REM following a NREM period automatically starts a potential REM period, however any REMP must be at least
## 5min (except the first REMP). NREMPs that exceed 120min in duration (excl. wake) can be split into 2 parts. 
## The new cycle then starts with the first N3 episode following a phase (>12min) with any other stage than N3 
## (cf. Rudzik et al., 2020; Jenni et al., 2004; Kurth et al., 2010).
##
## Besides sleep cycles (NREM-REM), the result also splits the NREM and REM parts of each cycle in percentiles. In case the 
## length of a period is not divisible by 10 (i.e., 203 epochs), we added one epoch to percentiles in a randomized
## fashion to reach the correct length of a period (i.e., 7 percentiles comprised 20 epochs, 3 comprised 21).
## 
## The code offers to choose whether incomplete cycles should be removed at the end of the night (rm_incompletecycs, default = F). 
##
## The user can either process all files in a given directory (default) or specific files by specifying a vector of files.
##
## By default, the function produces and saves a plot for visual inspection of the results.
################################################################################################################################## 
SleepCycles_CB <- function(p, files = NA, filetype = "vmrk", treat_as_W = NA, rm_incompletecycs = F, plot = T){
  
  # # --- set a few things
  # p <- "E:/10 Data/12-ALocGlo/rawdata_ALocGlo_oddball/LightSleep_EXP/SIESTA_results/bvaMarker_30s/" # path to SIESTA results converted to *vmrk using ascii2vmrk.m (located here: E:\10 Data\12-ALocGlo\rawdata_ALocGlo_oddball\LightSleep_EXP\SIESTA_results)
  setwd(p)
  
  #----- check if pacman is installed - if not install it
  if(!require(pacman)) install.packages("pacman")
  
  #----- use pacman function p_load to check all packages that you are using in this script
  pacman::p_load(ggplot2, reshape2, plyr, stringr, viridis)
  
  x <- list.files(p, pattern = glob2rx("*SCycles.txt"))
  if (length(x)>0){
    stop("Please remove files from previous Sleep Cycle detections from the folder.")
  }
  
  #----- list all files in folder
  if (filetype == "vmrk"){
    d <- list.files(p, pattern = "*.vmrk")
  }else if (filetype == "vpd"){
#################################################################
  }else if (filetype == "txt"){
    d <- list.files(p, pattern = "*.txt")
    hd <- readline("Do your files have a header with row names (y/n)? ") 
  }
  
  #----- has a vector of files been specified?
  if (!all(is.na(files))){
    d <- d[files]
  }
  
  #----- prepare results folder
  sv <- paste("SleepCycles", Sys.Date(), sep = "_")
  dir.create(file.path(paste(p, sv, sep = "/")), showWarnings = FALSE)
  
  #--------------------------------------------------------
  #----- loop through files to determine sleep cycles
  #--------------------------------------------------------
  
  for (i in 1:length(d)){
    print(i)
    filename <- d[i]
    
    ## load data
    if (filetype == "vmrk"){
      header <- read.csv(filename, nrows = 1, header = F)
      data <- read.csv(filename, skip = 1) #each sleep stage refers to the 30s preceding the marker (irrespective of SR!)
      cycles <- data
      cycles[,1] <- "SleepCycle"
      cycles[,2] <- NA
    }else if (filetype == "txt"){
      if (hd == "y"){ # does it have a header?
        data <- read.table(filename, header = T)
        for (z in 1:ncol(data)){
          if (length(unique(data[,z])) == 5){
            if (all(as.numeric(sort(unique(data[,z]))) == c(0,1,2,3,5))){
              colnames(data)[z] <- "Description"
              cycles <- data
              colnames(cycles)[z] <- "SleepCycle"
              break
            }
          }
          if (length(unique(data[,z])) == 6){
            if (all(as.numeric(sort(unique(data[,z]))) == na.omit(c(0,1,2,3,5,treat_as_W)))){
              colnames(data)[z] <- "Description"
              cycles <- data
              colnames(cycles)[z] <- "SleepCycle"
              break
            }
          }
        }
      }else{
        data <- read.table(filename, header = F)
        for (z in 1:ncol(data)){
          if (length(unique(data[,z])) == 5){
            if (all(as.numeric(sort(unique(data[,z]))) == c(0,1,2,3,5))){
              colnames(data)[z] <- "Description"
              cycles <- data
              colnames(cycles)[z] <- "SleepCycle"
              break
            }
          }
          if (length(unique(data[,z])) == 6){
            if (all(as.numeric(sort(unique(data[,z]))) == na.omit(c(0,1,2,3,5,treat_as_W)))){
              colnames(data)[z] <- "Description"
              cycles <- data
              colnames(cycles)[z] <- "SleepCycle"
              break
            }
          }
        }
      }
    }
    
    # Recode markers to be treated as W
    if (!is.na(treat_as_W)){
      data$Description[data$Description == treat_as_W] <- 0
    }
    
    # Recode/combine stages
    data$Descr2 <- NA
    data$Descr2[data$Description == 1 | data$Description == 2 | data$Description == 0 | data$Description == 5] <- "RWN12"
    data$Descr2[data$Description == 3] <- "N3"
    data$Descr3 <- NA
    data$Descr3[data$Description == 1 | data$Description == 2 | data$Description == 3] <- "NREM"
    data$Descr3[data$Description == 0] <- "W"
    data$Descr3[data$Description == 5] <- "REM"
    
    ## Find NREM periods: start with N1 and can then also include W. >=15min
    NREMWs <- which(data$Descr3 == "NREM"| data$Descr3 == "W") #which 30s epochs are NREM or wake
    NREMs <- which(data$Descr3 == "NREM")
    NREMWs <- subset(NREMWs, NREMWs >= NREMs[1]) # exclude W at the beginning of the night
    
    ## Loop through NREMWs 
    ## check if the sequence of NREWM is continuous and the period is >=15min AND beginning is not wake
    NREMWs_start <- NA
    for (k in 1:(length(NREMWs)-29)){
      if ((all(seq(NREMWs[k],length.out = 30) == NREMWs[seq(k,k+29)])) & (data$Descr3[NREMWs[k]] != "W")){ 
        NREMWs_start <- c(NREMWs_start, NREMWs[k])
      }else{
        next
      }
    }
    NREMWs_start <- NREMWs_start[-c(1)] #first was NA, remove
    
    # find discontinuities in the sequence (= potential beginnings of new NREM period)
    NREMWs_start2 <- NREMWs_start[1]
    for (k in 1:(length(NREMWs_start)-1)){
      if ((NREMWs_start[k+1]-NREMWs_start[k])>1){
        NREMWs_start2 <- c(NREMWs_start2, NREMWs_start[k+1]) #if there is a discontinuity in the sequence, mark the beginning of a new NREM period
      }
    }
    data$CycleStart <- NA
    data$CycleStart[NREMWs_start2] <- "NREMP"
    
    ## Find REM episodes (first can be <5min, others have to be at least 5min)
    REMs <- which(data$Descr3 == "REM") #which 30s epochs are NREM
    REMs_start <- REMs[1] #set beginning of first as no criterion
    for (k in 1:(length(REMs)-9)){
      if (all(seq(REMs[k],length.out = 10) == REMs[seq(k,k+9)])){ # check if the sequence of NREM epochs is continuous
        REMs_start <- c(REMs_start, REMs[k])
      }
    }
    REMs_start <- unique(REMs_start)
    
    REMs_start2 <- REMs_start[1] 
    for (k in 1:(length(REMs_start)-1)){
      if ((REMs_start[k+1]-REMs_start[k])>1){
        REMs_start2 <- c(REMs_start2, REMs_start[k+1]) #if there is an discontinuity in the sequence, mark the beginning of a new NREM sequence
      }
    }
    data$CycleStart[REMs_start2] <- "REMP"
    
    ## remove several NREMPs or REMPs in a row
    rm <- NA
    cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP")
    for (k in 2:length(cycs)){
      if(data$CycleStart[cycs[k]] == data$CycleStart[cycs[k-1]])
        rm <- c(rm, cycs[k])
    }
    rm <- unique(rm)
    rm <- rm[c(-1)]
    data$CycleStart[c(rm)] <- NA
    
    ## is any NREM part (excl. wake) of a NREMP longer than 120min?
    cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP")
    scndNREMP <- which(data$CycleStart == "NREMP")[2] 
    
    toolong <- NA
    for (k in seq(2,length(cycs),2)){ #check every second one as only every second is a NREMP
      subset <- data[c(cycs[k-1]:(cycs[k]-1)),]
      wake_eps <- sum(subset$Descr2 == "W")
      if (((cycs[k]-cycs[k-1])-wake_eps)>=240){ #<= as cycs[k] is already the beginning of the REMP
        toolong <- c(toolong, cycs[k-1])
      }
    }
    
    # now split the first NREMP that is too long
    if ((length(toolong) > 1) & (toolong[c(2)] < scndNREMP)){ #start of too long NREM must be before start of second NREMP -> only first NREMP is split
      toolong <- toolong[c(2)] #only choose first NREM cycle
      
      beg_end <- c(cycs[which(cycs==toolong)], cycs[which(cycs==toolong)+1]) #find beginning and end of NREMP that is too long
      
      # find RWN12 episodes that are > 12min
      RWN12s <- which(data$Descr2 == "RWN12") #which 30s epochs are N1/2
      RWN12s <- RWN12s[c(RWN12s>=beg_end[1] & RWN12s<=beg_end[2])]
      RWN12s_start <- NA
      for (kk in 1:(length(RWN12s)-23)){
        if (all(seq(RWN12s[kk],length.out = 24) == RWN12s[seq(kk,kk+23)])){ # check if the sequence of N12 epochs is continuous
          RWN12s_start <- c(RWN12s_start, RWN12s[kk])
        }
      }
      RWN12s_start <- RWN12s_start[-c(1)] #first was NA, remove
      rm(kk)
      
      RWN12s_start2 <- RWN12s_start[1]
      for (kk in 1:(length(RWN12s_start)-1)){
        if ((RWN12s_start[kk+1]-RWN12s_start[kk])>1){
          RWN12s_start2 <- c(RWN12s_start2, RWN12s_start[kk+1]) #if there is an discontinuity in the sequence, mark the beginning of a new NREM sequence
        }
      }
      
      # find first N3 episode
      N3s_1 <- which(data$Descr2 == "N3")[1] #which is the first 30s epoch of N3
      
      # delete RWN12s_start2 before first N3s
      RWN12s_start2 <- RWN12s_start2[c(RWN12s_start2>N3s_1)]
      
      # second NREMP starts with N3 following 12min of N1/2
      N3s <- which(data$Descr2 == "N3") #which 30s epoch of N3 are there
      N3s_cycstart <- N3s[c(N3s > RWN12s_start2[1])][1] # gives first N3 epoch after N12 episode >12min
      data$CycleStart[N3s_cycstart] <- "NREMP"
      
      splits <- N3s_cycstart
      splits <- unique(splits)
      
      # ask user if s/he is happy with the result of the splitting
      # plot results
      dfplot <- data
      dfplot$time <- seq(1, nrow(dfplot)) # gives epochs
      dfplot$Description[dfplot$Description == 1] <- -1
      dfplot$Description[dfplot$Description == 2] <- -2
      dfplot$Description[dfplot$Description == 3] <- -3
      dfplot$Description[dfplot$Description == 5] <- 1
      
      pp <- ggplot(dfplot, aes(x=time, y=Description, colour=Description)) 
      pp <- pp + theme_bw()+
        geom_point() +
        geom_line(aes(x=time, y=Description))+
        ggtitle(as.character(filename))+
        xlab("Time") +
        ylab("Sleep Stage")+
        scale_y_continuous(limits = c(-3,2), breaks = c(-3, -2, -1, 0, 1), labels = c("N3", "N2", "N1", "W", "REM"))+
        scale_color_viridis(name = "Sleep Stage", option = "D")+
        geom_vline(xintercept = c(splits), lty = 2, colour = "red")+
        annotate(geom="text", x = 500, y = 2, label = paste("split at epoch(s):", as.character(splits), sep = " "))
      print(pp)
      
      val <- readline("Are you happy with the result (y/n/skip)?. ") 
      
      if(val == "skip"){
        message("This night is skipped.")
        next
      }else if (val == "n"){
        newperiod <- readline("At which epoch do you want to start the new NREM period instead? Please type epoch number or NA to not split. ") 
        data$CycleStart[N3s_cycstart] <- NA #rm old indicator
        if (newperiod != "NA"){
          data$CycleStart[as.numeric(newperiod)] <- "NREMP"
        }
      }else if (val == "y"){
        message("Yay, that seemed to work well.")
      }else{
        message("Missing entry. This night is skipped")
      }
      rm(dfplot, pp)
    }
    
    
    # -----------------------------------------------------------------------------
    # now finish and add cycle markers and percentiles
    
    # add cycle markers to file (only for NREM cycles)
    NREMPs <- which(data$CycleStart == "NREMP") #where do NREMPs begin?
    
    data$cycles <- NA
    for (k in 1:(length(NREMPs))){
      if (k<length(NREMPs)){
        data$cycles[NREMPs[k]:(NREMPs[k+1]-1)] <- k 
      }else{
        data$cycles[NREMPs[k]:(nrow(data))] <- k 
      }
    }
    
    # add NREM vs. REM part info
    cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP")
    data$REM.NREM <- NA
    for (k in 1:(length(cycs))){
      if (k<length(NREMPs)){
        if (data$CycleStart[cycs[k]] == "NREMP"){ 
          data$REM.NREM[c((cycs[k]):((cycs[k+1])-1))] <- 0
        }else if (data$CycleStart[cycs[k]] == "REMP"){
          data$REM.NREM[c((cycs[k]):((cycs[k+1])-1))] <- 1
        }
      }else{
        if (data$CycleStart[cycs[k]] == "NREMP"){ 
          data$REM.NREM[c((cycs[k]):(nrow(data)))] <- 0
        }else if (data$CycleStart[cycs[k]] == "REMP"){
          data$REM.NREM[c((cycs[k]):(nrow(data)))] <- 1
        }
      }
    }
    
    # remove incomplete NREM-REM cycle at the end of the night (i.e., cycles followed by <5min NREM or W)
    if(!exists("rm_incompletecycs")){
      stop ("Please specify whether you want to remove or keep the incomplete last cycle.")
    } 
    
    if (rm_incompletecycs == T){
      cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP") #remove NREM/W following last REMP or REM/W following last NREMP
      if (data$CycleStart[cycs[length(cycs)]] == "REMP"){
        REMs <- which(data$Descr3 == "REM") #which 30s epochs are REM
        stop <- REMs[length(REMs)]+1 # stop after last REM epoch
        end <- which(data$Descr3 == "NREM")
        end <- end[end>=stop]
        if(length(end)>10){
          data$CycleStart[stop] <- "stop"
          data$cycles[stop:nrow(data)] <- NA
          data$REM.NREM[stop:nrow(data)] <- NA
        }else{
          # remove last REMP
          stop <- cycs[length(cycs)]
          data$CycleStart[cycs[length(cycs)]] <- "stop"
          data$cycles[stop:nrow(data)] <- NA
          data$REM.NREM[stop:nrow(data)] <- NA
        }
      }else{
        NREMs <- which(data$Descr3 == "NREM") #which 30s epochs are NREM
        stop <- NREMs[length(NREMs)]+1 # stop after last NREM epoch
        end <- which(data$Descr3 == "REM")
        end <- end[end>=stop]
        if(length(end)>10){
          data$CycleStart[stop] <- "stop"
          data$cycles[stop:nrow(data)] <- NA
          data$REM.NREM[stop:nrow(data)] <- NA
        }else{
          # remove last NREMP
          stop <- cycs[length(cycs)]
          data$CycleStart[cycs[length(cycs)]] <- "stop"
          data$cycles[stop:nrow(data)] <- NA
          data$REM.NREM[stop:nrow(data)] <- NA
        }
      }
      lastREMP <- cycs[length(cycs)] # find end of last REMP
      end <- which(data$Descr3 == "NREM" | data$Descr3 == "W")
      end <- end[c(end > lastREMP)] # NREM or wake following onset of last REMP
      stop <- NA
    }else if (rm_incompletecycs == F){
      #remove NREM/W following last REMP or REM/W following last NREMP
      cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP") 
      if (data$CycleStart[cycs[length(cycs)]] == "REMP"){
        REMs <- which(data$Descr3 == "REM") #which 30s epochs are REM
        if((length(REMs) + 1)>length(REMs)){ #check if the last REM is the last staged epoch
          stop <- REMs[length(REMs)] # stop at last (REM) epoch
          data$CycleStart[stop] <- "stop"
        }else{
          stop <- REMs[length(REMs)]+1 # stop after last REM epoch
          data$CycleStart[stop] <- "stop"
          data$cycles[stop:nrow(data)] <- NA
          data$REM.NREM[stop:nrow(data)] <- NA
        }
      }else{
        lastNREMP <-  tail(which(data$CycleStart == "NREMP"),1)
        Ws <- which(data$Descr3 == "W") #which 30s epochs are W
        Ws <- Ws[Ws > lastNREMP]
        if (length(Ws) > 2){
          # find last continuous W episode (i.e., final W)
          Ws_start <- NA
          for (kk in 2:(length(Ws))){
            if (Ws[kk-1] != Ws[kk]-1){ 
              Ws_start <- c(Ws_start, Ws[kk])
            }else{
              next
            }
          }
          if (!all(is.na(Ws_start))){
            Ws_start <- tail(Ws_start,1) #last is the onset of the last wake episode
          }else{
            Ws_start <- head(Ws,1)
          }
          data$CycleStart[Ws_start] <- "stop"
          # now check again if >=15min criterion is still fulfilled for the last NREMP
          if(Ws_start - tail(which(data$CycleStart == "NREMP"),1)>=30){
            data$cycles[Ws_start:nrow(data)] <- NA
            data$REM.NREM[Ws_start:nrow(data)] <- NA
          }else{
            # if not, remove last NREMP completely
            data$CycleStart[tail(which(data$CycleStart == "NREMP"),1)] <- "stop"
            data$CycleStart[Ws_start] <- NA
            data$cycles[tail(which(data$CycleStart == "stop"),1):nrow(data)] <- NA
            data$REM.NREM[tail(which(data$CycleStart == "stop"),1):nrow(data)] <- NA
            
            #now again remove NREM/W following last REMP
            cycs <- which(data$CycleStart == "NREMP" | data$CycleStart == "REMP") 
            if (data$CycleStart[cycs[length(cycs)]] == "REMP"){
              REMs <- which(data$Descr3 == "REM") #which 30s epochs are REM
              stop <- REMs[length(REMs)]+1 # stop after last REM epoch
              data$CycleStart[stop] <- "stop"
              data$cycles[stop:nrow(data)] <- NA
              data$REM.NREM[stop:nrow(data)] <- NA
            }
          }
        }
      }
    }
    
    ## merge cycle marker & NREM/REM marker
    data$CycInfo <- paste(data$cycles, data$REM.NREM, sep = "")
    data$CycInfo[data$CycInfo == "NANA"] <- NA
    
    ## add percentiles of NREM & REM parts
    cycs <- na.omit(unique(data$CycInfo))
    data$perc <- NA
    for (k in 1:length(cycs)){
      length <- nrow(subset(data, data$CycInfo == cycs[k]))
      if ((length %% 10) == 0){
        seq <- rep(length/10,10)
        for (kk in 1:length(seq)){
          if (kk == 1){
            ind <- rep(kk, seq[kk])
          }else{
            ind <- c(ind, rep(kk, seq[kk]))
          }
        }
        data$perc[which(data$CycInfo == cycs[k])] <- ind
      }else{
        seq <- sample(c(rep(floor(length/10), (10-length %% 10)), rep(ceiling(length/10), length %% 10))) #construct shuffled sequence of floor/ceiling rounded percentiles that matches the length of the complete cycle
        for (kk in 1:length(seq)){
          if (kk == 1){
            ind <- rep(kk, seq[kk])
          }else{
            ind <- c(ind, rep(kk, seq[kk]))
          }
        }
        data$perc[which(data$CycInfo == cycs[k])] <- ind
      }
    }
    
    ## merge cycle marker, NREM/REM marker, & percentile marker
    data$cycle_perc <- paste(data$CycInfo, data$perc, sep = "_")
    data$cycle_perc[data$cycle_perc == "NA_NA"] <- NA
    
    ## prep new marker file with cycle info
    cycles <- cycles[,-c(4,5)]
    cycles$Description <- data$cycle_perc
    cycles$SleepCycle <- data$cycles
    cycles$N_REM <- data$REM.NREM
    cycles$percentile <- data$perc
    
    ## save new file
    svv <- paste(p, sv, sep = "/")
    name <- unlist(str_split(filename, pattern = "_"))
    savename <- paste0(c(name[1:(length(name)-1)]), sep = "_", collapse = "")
    savename <- paste(savename, "SCycles.txt", sep = "")
    write.table(cycles, file = paste(svv, savename, sep = "/"), row.names = F)
    
    ## plot results if desired
    if (plot == T){
      dfplot <- data
      dfplot$time <- seq(1,nrow(dfplot))
      if (filetype == "vmrk"){
        dfplot$time2 <- (dfplot$Position/(1000/128))/(1000*60)
      }
      dfplot$Description[dfplot$Description == 1] <- -1
      dfplot$Description[dfplot$Description == 2] <- -2
      dfplot$Description[dfplot$Description == 3] <- -3
      dfplot$Description[dfplot$Description == 5] <- 1
      dfplot$CycInfo[dfplot$CycInfo == 10] <- 1.5
      dfplot$CycInfo[dfplot$CycInfo == 11] <- 1.6
      dfplot$CycInfo[dfplot$CycInfo == 20] <- 1.8
      dfplot$CycInfo[dfplot$CycInfo == 21] <- 1.9
      dfplot$CycInfo[dfplot$CycInfo == 30] <- 1.5
      dfplot$CycInfo[dfplot$CycInfo == 31] <- 1.6
      dfplot$CycInfo[dfplot$CycInfo == 40] <- 1.8
      dfplot$CycInfo[dfplot$CycInfo == 41] <- 1.9
      dfplot$CycInfo[dfplot$CycInfo == 50] <- 1.5
      dfplot$CycInfo[dfplot$CycInfo == 51] <- 1.6
      dfplot$CycInfo[dfplot$CycInfo == 60] <- 1.8
      dfplot$CycInfo[dfplot$CycInfo == 61] <- 1.9
      dfplot$CycInfo[dfplot$CycInfo == 70] <- 1.5
      dfplot$CycInfo[dfplot$CycInfo == 71] <- 1.6
      dfplot$CycInfo <- as.numeric(dfplot$CycInfo)
      
      ggplot(dfplot, aes(x=time, y=Description, colour=Description)) + 
        theme_bw()+
        geom_point() +
        geom_line(aes(x=time, y=Description))+
        xlab("Time") +
        ylab("Sleep Stage")+
        # scale_y_continuous(limits = c(-3,5), labels = c("N3", "N2", "N1", "W", "REM", "", "", "", ""))+
        scale_y_continuous(limits = c(-3,2), breaks = c(-3, -2, -1, 0, 1, 2, 3, 4, 5), labels = c("N3", "N2", "N1", "W", "REM", "", "", "", ""))+
        scale_color_viridis(name = "Sleep Stage", option = "D")+
        geom_point(y = dfplot$CycInfo, size = 0.5)
      
      savename <- paste0(c(name[1:(length(name)-2)]), sep = "_", collapse = "")
      savename <- paste(savename, "plot.png", sep = "")
      ggsave(file=paste(svv, savename, sep = "/"), width = 25, height = 15, units = "cm", dpi = 600) 
    }
  }
}
